File: Data Structure+Algorithm/Sorting/Merge Sort/mergesort.cpp
#include<bits/stdc++.h>
using namespace std;
void merge(int arr[],int l,int m,int r)    //function to merge two sorted arrays
{
	int n1 = m-l+1;
	int n2 = r-m;
	int L[n1],R[n2];        // creating left and right sub-array

	for(int i=0;i<n1;i++){
		L[i] = arr[l+i];
	}
	for(int i=0;i<n2;i++){
		R[i] = arr[m+i+1];
	}
	int i=0,j=0,k=l;
	while(i<n1 && j<n2){
		if(L[i]<=R[j]){
			arr[k++]=L[i++];
		}
		else{
			arr[k++]=R[j++];
		}
	}
	while(i<n1){
		arr[k++]=L[i++];
	}
	while(j<n2){
		arr[k++]=R[j++];
	}
}

void merge_sort(int arr[],int l,int r)   // Recursive function to perform merge sort
{
	if(l < r)
	{
		int m = l+(r-l)/2;
		merge_sort(arr,l,m);
		merge_sort(arr,m+1,r);
		merge(arr,l,m,r);
	}
}

void mergeSort(int n,int arr[])
{
	merge_sort(arr,0,n);
}

int main() 
{
	int n = 7;
	int arr[7] = {9, 13, 4, 6, 2, 11, 5};
	mergeSort(n, arr);

	for(int i=0; i<n; i++)   //function to print an sorted array
	{
		cout<<arr[i]<<endl;
	}
	return 0;
}

File: Data Structure+Algorithm/Sorting/Advanced/Merge Sort/merge_sort.js
/**
 * Merge - To merge two sorted array
 * @param {array} arr1 
 * @param {array} arr2 
 * @return Array
 */
function merge(arr1, arr2) {
    let result = []
    let i = 0
    let j = 0
    while (i < arr1.length && j < arr2.length) {
        if (arr1[i] < arr2[j]) {
            result.push(arr1[i])
            i++
        } else {
            result.push(arr2[j])
            j++
        }
    }
    while (i < arr1.length) {
        result.push(arr1[i])
        i++
    }
    while (j < arr2.length) {
        result.push(arr2[j])
        j++
    }
    return result
}
/**
    mergeSort - to sort given array
    @param {arr} array - Array to sort
*/
function mergeSort(arr) {
    if (arr.length <= 1) return arr
    let mid = Math.floor(arr.length / 2)
    let left = mergeSort(arr.slice(0, mid))
    let right = mergeSort(arr.slice(mid))
    return merge(left, right)
}
console.log(mergeSort([43, 56, 33, 1, 22, 67, 45]))